src/Components/Components/src/Routing/UrlValueConstraint.cs

```diff
@@ -0,0 +1,176 @@
+
+        private static bool TryParse(ReadOnlySpan<char> str, out double result)
+            => double.TryParse(str, NumberStyles.Number, CultureInfo.InvariantCulture, out result);
+
+        private static bool TryParse(ReadOnlySpan<char> str, out float result)
+            => float.TryParse(str, NumberStyles.Number, CultureInfo.InvariantCulture, out result);
+
+        private static bool TryParse(ReadOnlySpan<char> str, out int result)
+            => int.TryParse(str, NumberStyles.Integer, CultureInfo.InvariantCulture, out result);
+
+        private static bool TryParse(ReadOnlySpan<char> str, out long result)
+            => long.TryParse(str, NumberStyles.Integer, CultureInfo.InvariantCulture, out result);
+
+        private static UrlValueConstraint? Create(Type targetType) => targetType switch
+        {
+            var x when x == typeof(string) => new TypedUrlValueConstraint<string>(TryParse),
+            var x when x == typeof(bool) => new TypedUrlValueConstraint<bool>(bool.TryParse),
+            var x when x == typeof(bool?) => new NullableTypedUrlValueConstraint<bool>(bool.TryParse),
+            var x when x == typeof(DateTime) => new TypedUrlValueConstraint<DateTime>(TryParse),
+            var x when x == typeof(DateTime?) => new NullableTypedUrlValueConstraint<DateTime>(TryParse),
+            var x when x == typeof(decimal) => new TypedUrlValueConstraint<decimal>(TryParse),
+            var x when x == typeof(decimal?) => new NullableTypedUrlValueConstraint<decimal>(TryParse),
+            var x when x == typeof(double) => new TypedUrlValueConstraint<double>(TryParse),
+            var x when x == typeof(double?) => new NullableTypedUrlValueConstraint<double>(TryParse),
+            var x when x == typeof(float) => new TypedUrlValueConstraint<float>(TryParse),
+            var x when x == typeof(float?) => new NullableTypedUrlValueConstraint<float>(TryParse),
+            var x when x == typeof(Guid) => new TypedUrlValueConstraint<Guid>(Guid.TryParse),
+            var x when x == typeof(Guid?) => new NullableTypedUrlValueConstraint<Guid>(Guid.TryParse),
+            var x when x == typeof(int) => new TypedUrlValueConstraint<int>(TryParse),
+            var x when x == typeof(int?) => new NullableTypedUrlValueConstraint<int>(TryParse),
+            var x when x == typeof(long) => new TypedUrlValueConstraint<long>(TryParse),
+            var x when x == typeof(long?) => new NullableTypedUrlValueConstraint<long>(TryParse),
+            var x => null
+        };
+
+        public abstract bool TryParse(ReadOnlySpan<char> value, [MaybeNullWhen(false)] out object result);
+
+        public abstract object? Parse(ReadOnlySpan<char> value, string destinationNameForMessage);
+
+        public abstract Array ParseMultiple(StringSegmentAccumulator values, string destinationNameForMessage);
+
+        private class TypedUrlValueConstraint<T> : UrlValueConstraint
```

Comment:
```suggestion
        private sealed class TypedUrlValueConstraint<T> : UrlValueConstraint
```

Reply:
Can't be because `NullableTypedUrlValueConstraint` inherits from it.

Reply:
I mean technically I could implement `NullableTypedUrlValueConstraint` in a different and independent way to avoid this, but I'm going to argue it's not worth it.